# STAT 207 Lab Assignment 9 - Individual Part


# <u>Case Study</u>: North American Video Game Sales *for New Video Games*


## Dataset

The data in the `video_game_sales_sample.csv` file contains information about a sample of video games released between 1995 and 2016.

* `NA_Sales`: yearly total North American sales (in millions of dollars) of the video game
* `EU_Sales`: yearly total European sales (in millions of dollars) of the video game
* `JP_Sales`: yearly total Japanese sales (in millions of dollars) of the video game
* `Other_Sales`: yearly total sales in other countries (in millions of dollars) of the video game
* `Year`: the year that video game was released
* `Genre`: (just considering action or strategy in this analysis) genre of the video game
* `Platform`: (just considering PS, PS2, PS3, DS, and Wii in this analysis) platform of the video game

## Research Goals

We would like to pursue the following two research goals in this analysis.

### Primary Research Goal

The primary research goal that we will pursue in this lab will be to build a predictive model that effectively predicts the `EU_Sales` (ie. the yearly total European sales, in millions of dollars) of *new video games* given some combination of the remaining variables:

* `NA_Sales`
* `JP_Sales`
* `Other_Sales`
* `Year`
* `Genre`
* `Platform`

### Secondary Research Goal

Ideally, the model that we select would also be interpretable.

## Points

<p>&nbsp;</p>

<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;width:51pt;">Problem</td>
            <td style="color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid black;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">1.7</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">3.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">3.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">3.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.5</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">4.6</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">5.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">5.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">5.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">5.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">6.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1.25</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">6.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">6.3</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">6.4</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">7.1</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;height:14.25pt;border-top:none;">7.2</td>
            <td style="color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid black;border-top:none;border-left:none;">1.25</td>
        </tr>
    </tbody>
</table>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import cross_val_score

from sklearn.linear_model import Ridge, Lasso, ElasticNet, LinearRegression

from sklearn.preprocessing import StandardScaler

import warnings
from sklearn.exceptions import ConvergenceWarning

# Ignore scikit-learn ConvergenceWarnings
warnings.filterwarnings("ignore", category=ConvergenceWarning)
warnings.filterwarnings("ignore", category=UserWarning)

## 1. Basic Data Examination and Cleaning

### 1.1. Reading the Dataset and Detecing Any Implicit Missing Values

1. **Reading and Implicit Missing Value Conversion**: Read this csv file into a dataframe. This csv file may contains implicit missing values in our intended numerical variables! Make sure that when you read your csv file into your dataframe (the final time), that any implicit missing values in the numerical variable columns have been converted into NaN values.

2. **Display your dataframe.**

3. **Show how many rows your dataframe has.**



In [2]:
df = pd.read_csv("video_game_sales_sample.csv")
df.dtypes

Name            object
Platform        object
Year           float64
Genre           object
NA_Sales        object
EU_Sales       float64
JP_Sales       float64
Other_Sales    float64
dtype: object

In [3]:
np.unique(df["NA_Sales"])

array(['0', '0.01', '0.02', '0.03', '0.04', '0.05', '0.06', '0.07',
       '0.08', '0.09', '0.1', '0.11', '0.12', '0.13', '0.14', '0.15',
       '0.16', '0.17', '0.18', '0.19', '0.2', '0.21', '0.22', '0.23',
       '0.24', '0.25', '0.26', '0.27', '0.28', '0.29', '0.3', '0.31',
       '0.32', '0.33', '0.34', '0.35', '0.36', '0.37', '0.38', '0.39',
       '0.4', '0.41', '0.42', '0.43', '0.44', '0.45', '0.46', '0.47',
       '0.48', '0.5', '0.51', '0.52', '0.53', '0.54', '0.55', '0.56',
       '0.57', '0.58', '0.59', '0.6', '0.61', '0.63', '0.64', '0.65',
       '0.66', '0.67', '0.69', '0.71', '0.72', '0.73', '0.75', '0.76',
       '0.78', '0.79', '0.8', '0.81', '0.82', '0.85', '0.86', '0.87',
       '0.88', '0.93', '0.96', '0.99', '1.01', '1.05', '1.06', '1.07',
       '1.12', '1.15', '1.21', '1.23', '1.27', '1.29', '1.3', '1.34',
       '1.36', '1.4', '1.41', '1.42', '1.43', '1.46', '1.48', '1.54',
       '1.56', '1.59', '1.7', '1.78', '1.79', '1.82', '1.88', '1.89',
       '1.9', '1.96

In [4]:
df = pd.read_csv("video_game_sales_sample.csv", na_values=["unknown"])
df.dtypes

Name            object
Platform        object
Year           float64
Genre           object
NA_Sales       float64
EU_Sales       float64
JP_Sales       float64
Other_Sales    float64
dtype: object

In [5]:
df.head()

Name Platform    Year     Genre  NA_Sales  \
0                  Saints Row IV      PS3  2013.0    Action      0.56   
1          Treasures of the Deep       PS  1997.0    Action      0.10   
2  Skylanders: Spyro's Adventure      Wii  2011.0    Action      1.40   
3     LEGO Batman: The Videogame      PS2  2008.0    Action      0.72   
4                  Carnage Heart       PS  1995.0  Strategy      0.01   

   EU_Sales  JP_Sales  Other_Sales  
0      0.44      0.09         0.21  
1      0.07      0.00         0.01  
2      1.14      0.00         0.31  
3      0.03      0.00         0.52  
4      0.01      0.09         0.01

In [6]:
shape1 = df.shape[0]
print(f"The dataframe has {shape1} rows")

The dataframe has 1034 rows


### 1.2. Dropping Missing Values

1. Drop all rows in this dataframe with a missing value.
2. Reset the index (we'll need a fresh index for what we do in 1.7).
3. How many rows did you drop?

In [7]:
df = df.dropna().reset_index()

In [8]:
shape2 = df.shape[0]
print(f"{shape1 - shape2} rows were dropped")

23 rows were dropped


### 1.3. Dropping Unneeded Variables

We don't intend to use the `Name` column in this analysis, so let's drop this variable from the dataframe.

In [12]:
df = df[["Platform","Year","Genre","NA_Sales","EU_Sales","JP_Sales","Other_Sales"]]
df

Platform    Year     Genre  NA_Sales  EU_Sales  JP_Sales  Other_Sales
0         PS3  2013.0    Action      0.56      0.44      0.09         0.21
1          PS  1997.0    Action      0.10      0.07      0.00         0.01
2         Wii  2011.0    Action      1.40      1.14      0.00         0.31
3         PS2  2008.0    Action      0.72      0.03      0.00         0.52
4          PS  1995.0  Strategy      0.01      0.01      0.09         0.01
...       ...     ...       ...       ...       ...       ...          ...
1006      PS2  2004.0    Action      0.37      0.47      0.54         0.07
1007       DS  2007.0  Strategy      0.00      0.00      0.26         0.00
1008      Wii  2007.0  Strategy      0.22      0.03      0.08         0.02
1009      PS3  2011.0    Action      0.30      0.23      0.00         0.09
1010       PS  1998.0    Action      0.01      0.01      0.00         0.00

[1011 rows x 7 columns]

### 1.4. Features Matrix and Target Array 

Use this full dataset (ie. not one that has been split into training or test dataset) to create a features matrix and target array for our linear regression models. 

In [10]:
df_features = df.drop(["EU_Sales"], axis=1)

In [11]:
df_target = df["EU_Sales"]

### 1.5. Scaling the Numerical Explanatory Variables

First create another dataframe that is comprised of just the numerical variables that have been z-score scaled.

In [17]:
scaler = StandardScaler()
scaled_features = scaler.fit_transform(df_features[["Year", "NA_Sales", "JP_Sales", "Other_Sales"]])
df_features_scaled = pd.DataFrame(scaled_features, columns=df_features[["Year", "NA_Sales", "JP_Sales", "Other_Sales"]].columns)
df_features_scaled

Year  NA_Sales  JP_Sales  Other_Sales
0     1.327470  0.547725  0.155697     0.679615
1    -2.132036 -0.300538 -0.377494    -0.287438
2     0.895032  2.096726 -0.377494     1.163142
3     0.246374  0.842772 -0.377494     2.178547
4    -2.564475 -0.466502  0.155697    -0.287438
...        ...       ...       ...          ...
1006 -0.618502  0.197355  2.821650     0.002678
1007  0.030155 -0.484943  1.162835    -0.335790
1008  0.030155 -0.079252  0.096454    -0.239085
1009  0.895032  0.068272 -0.377494     0.099384
1010 -1.915817 -0.466502 -0.377494    -0.335790

[1011 rows x 4 columns]

### 1.6. Creating Indicator Variables

Next, unfortunately our scikit-learn regression equations require us to creat our own 0/1 indicator variables for each of our categorical explanatory variables.

You can do this by using the **pd.get_dummies()** function, specifically setting **drop_first=True** and **dtype=int**.

`pd.get_dummies(INSERT_YOUR_DATAFRAME_WITH_JUST_CAT_VARIABLES, drop_first=True, dtype=int)`


Create this dataframe of the appropriate 0/1 indicator variables below.

In [20]:
df_features_dummies = pd.get_dummies(df_features[["Platform", "Genre"]], drop_first=True, dtype=int)
df_features_dummies

Platform_PS  Platform_PS2  Platform_PS3  Platform_Wii  Genre_Strategy
0               0             0             1             0               0
1               1             0             0             0               0
2               0             0             0             1               0
3               0             1             0             0               0
4               1             0             0             0               1
...           ...           ...           ...           ...             ...
1006            0             1             0             0               0
1007            0             0             0             0               1
1008            0             0             0             1               1
1009            0             0             1             0               0
1010            1             0             0             0               0

[1011 rows x 5 columns]

### 1.7. New Features Matrix

Overwrite your features matrix to be the **horizontal concatenation** of your two dataframes from 1.5 and 1.6. You can do this by using the **pd.concat()** function

`pd.concat([1.5_DATAFRAME, 1.6_DATAFRAME], axis=1)`

In [21]:
df_features = pd.concat([df_features_scaled, df_features_dummies], axis=1)
df_features

Year  NA_Sales  JP_Sales  Other_Sales  Platform_PS  Platform_PS2  \
0     1.327470  0.547725  0.155697     0.679615            0             0   
1    -2.132036 -0.300538 -0.377494    -0.287438            1             0   
2     0.895032  2.096726 -0.377494     1.163142            0             0   
3     0.246374  0.842772 -0.377494     2.178547            0             1   
4    -2.564475 -0.466502  0.155697    -0.287438            1             0   
...        ...       ...       ...          ...          ...           ...   
1006 -0.618502  0.197355  2.821650     0.002678            0             1   
1007  0.030155 -0.484943  1.162835    -0.335790            0             0   
1008  0.030155 -0.079252  0.096454    -0.239085            0             0   
1009  0.895032  0.068272 -0.377494     0.099384            0             0   
1010 -1.915817 -0.466502 -0.377494    -0.335790            1             0   

      Platform_PS3  Platform_Wii  Genre_Strategy  
0                1             0               0  
1                0             0               0  
2                0             1               0  
3                0             0               0  
4                0             0               1  
...            ...           ...             ...  
1006             0             0               0  
1007             0             0               1  
1008             0             1               1  
1009             1             0               0  
1010             0             0               0  

[1011 rows x 9 columns]

## 2. Preliminary Analysis

We intend to build linear regression models that predict `EU_Sales` with some of the following explanatory variables.

* `NA_Sales`
* `JP_Sales`
* `Other_Sales`
* `Year`
* `Genre`
* `Platform`

Will a non-regularized linear regression model that uses ALL of these explanatory variables have an issue with multicollinearity? Explain.

In [22]:
df_features.corr()

Year  NA_Sales  JP_Sales  Other_Sales  Platform_PS  \
Year            1.000000 -0.070740 -0.117194     0.037612    -0.756297   
NA_Sales       -0.070740  1.000000  0.297544     0.743305     0.028450   
JP_Sales       -0.117194  0.297544  1.000000     0.289820     0.087765   
Other_Sales     0.037612  0.743305  0.289820     1.000000    -0.059531   
Platform_PS    -0.756297  0.028450  0.087765    -0.059531     1.000000   
Platform_PS2   -0.235704  0.027230  0.062344     0.084224    -0.231725   
Platform_PS3    0.508411  0.044864  0.048030     0.138741    -0.227379   
Platform_Wii    0.212839 -0.022556 -0.090327    -0.072822    -0.169438   
Genre_Strategy -0.225630 -0.150113  0.059801    -0.115321     0.200339   

                Platform_PS2  Platform_PS3  Platform_Wii  Genre_Strategy  
Year               -0.235704      0.508411      0.212839       -0.225630  
NA_Sales            0.027230      0.044864     -0.022556       -0.150113  
JP_Sales            0.062344      0.048030     -0.090327        0.059801  
Other_Sales         0.084224      0.138741     -0.072822       -0.115321  
Platform_PS        -0.231725     -0.227379     -0.169438        0.200339  
Platform_PS2        1.000000     -0.317234     -0.236397        0.004086  
Platform_PS3       -0.317234      1.000000     -0.231963       -0.155779  
Platform_Wii       -0.236397     -0.231963      1.000000       -0.031884  
Genre_Strategy      0.004086     -0.155779     -0.031884        1.000000

The only instances of multicollinearity occur between `Platfom_PS` and `Year` and between `NA_Sales` and `Other_Sales`.

## 3. Naive Implementation of a LASSO Model Cross-Validation Analysis

In #4 we will conduct parameter tuning on the $\lambda$ value in a LASSO linear regression model. We will use k=5 fold cross-validation to help us select the best $\lambda$ value for a LASSO model. But for now, let's start with the basics and perform k=5 fold cross-validation on just a single LASSO model with $\lambda=0.25$.

### 3.1. Instantiate the LASSO Model

First, instanatiate (but don't fit) a LASSO model object with $\lambda=0.25$.

### 3.2. k=5 Fold Cross-Validation

Next, instantiate a k=5 fold object that *will* randomly split the dataset up into k=5 test folds. Your dataset should be randomly shuffled first before creating the test folds. Use a random state of `207`.

### 3.3. Cross-Validation

1. Finally, use these two objects that you created in 3.1 and 3.2 to perform k=5 fold cross-validation on your features matrix from 1.7 and your target array. You should use the *R^2* to evaluate each of your 5 training dataset model's corresponding test datasets.
2. Display each of the 5 test R^2 values.
3. Display the mean of these 5 test R^2 values.
4. Display the standard deviation of these 5 test R^2 values.

## 4. Parameter Tuning of a LASSO Model with Cross-Validation

Next, let's use the code that we used in 3.1, 3.2, and 3.3 to test the performance of MANY LASSO models with different values of $\lambda$. Specifically, let's try to find the $\lambda$ value in a LASSO model that yields the highest *average* test R^2 value in k=5 fold cross-validation.


### 4.1. Parameter Tuning 

1. In a for loop, iterate through a series of $\lambda$ values that go from [0,0.01, 0.02, 0.03,...,0.6].
2. For each value of $\lambda$ do the following.
    * Perform k=5 fold cross-validation of the LASSO linear regression model with the features matrix (1.7) and target array and this given $\lambda$.
    * Calculate the **average** test R^2 value of each of your resulting 5 trained models.
    * Calculate the **standard deviation** test R^2 value of each of your resulting 5 trained models.
    
Your final result of this problem should include a *dataframe* that has three columns:
* the $\lambda$ value
* the *average* test R^2 that correspond to this $\lambda$ value
* the *standard deviation* test R^2 that correspond to this $\lambda$ value.

### 4.2. Average Test R^2 and $\lambda$ Relationship

Plot the relationship between the $\lambda$ values and the average test R^2 in a **line plot**.

### 4.3. Highest Average Test R^2

Which of our LASSO linear regression models do you think might perform the best when predicting `EU_Sales` for *new video games*?

### 4.4. Fitting the Best LASSO Model

Fit this best LASSO model that you selected in 4.3 with your features matrix (from 1.7) and your target array. Then, write out the linear regression equation that corresponds to this fitted model. Make sure that you use the appropriate notation discussed in class.

Furthermore, if any of your models slopes have been zeroed out, do not include the corresponding explanatory variables in the model.

### 4.5. Interpreting the LASSO Model


Were there any explanatory variables that our cross-validation suggested may have been overfitting our model? If so, which ones?

### 4.6. Multicollinearity

Do any of the explanatory variables that correspond to slopes in this best LASSO model that have not been zeroed out exhibit a strong linear relationship with each other?

## 5. Non-Regularized Linear Regression Cross-Validation Analyses

### 5.1. Full Model

First, perform k=5 fold cross-validation analysis on a non-regularized linear regression model using the features matrix (from 1.7) and target array.

Make sure you use the same 5 test folds that you created and used in #3 and #4.

You should use the *R^2* to evaluate each of your 5 training dataset model's corresponding test datasets.
1. Display each of the 5 test R^2 values.
2. Display the mean of these 5 test R^2 values.
3. Display the standard deviation of these 5 test R^2 values.

### 5.2. Reduced Model

Next, perform k=5 fold cross-validation analysis on a non-regularized linear regression model using **only the columns in the features matrix (from 1.7) that had non-zero slopes in 4.4** and the target array.

Make sure you use the same 5 test folds that you created and used in #3 and #4.

You should use the *R^2* to evaluate each of your 5 training dataset model's corresponding test datasets.
1. Display each of the 5 test R^2 values.
2. Display the mean of these 5 test R^2 values.
3. Display the standard deviation of these 5 test R^2 values.

### 5.3. Fitting the Reduced Model

Fit an ACTUAL non-regularized linear regression model using **only the columns in the features matrix (from 1.7) that had non-zero slopes in 4.4** and your target array. Then, write our the linear regression equation that corresponds to this fitted model. Make sure that you use the appropriate notation discussed in class.


### 5.4. Comparing Model Performance

Which model had better average test dataset performance: 
1. the best LASSO model (4.3, 4.4) or 
2. the non-regularized reduced model (5.2,5.3) *(which we built based on the best LASSO model's feature selection suggestions*)?

## 6. Parameter Tuning of a Ridge Regression Model with Cross-Validation

Next, let's test the performance of MANY ridge regression models with different values of $\lambda$. Specifically, let's try to find the $\lambda$ value in a ridge regression model that yields the highest *average* test R^2 value in k=5 fold cross-validation.


### 6.1. Parameter Tuning 

1. In a for loop, iterate through a series of $\lambda$ values that go from [0,1,2,...,450].
2. For each value of $\lambda$ do the following.
    * Perform k=5 fold cross-validation of the ridge regression model with the features matrix (1.7) and target array and this given $\lambda$.
    * Calculate the **average** test R^2 value of each of your resulting 5 trained models.
    * Calculate the **standard deviation** test R^2 value of each of your resulting 5 trained models.
    
Your final result of this problem should include a *dataframe* that has three columns:
* the $\lambda$ value
* the *average* test R^2 that correspond to this $\lambda$ value
* the *standard deviation* test R^2 that correspond to this $\lambda$ value.

### 6.2. Average Test R^2 and $\lambda$ Relationship

Plot the relationship between the $\lambda$ values and the average test R^2 in a **line plot**.

### 6.3. Highest Average Test R^2

Which of our fitted ridge regression models do you think might perform the best when predicting `EU_Sales` for *new video games*?

### 6.4. Fitting the Best Ridge Regression Model

Fit this best ridge regression model that you selected in 6.3 with your features matrix (from 1.7) and your target array. Then, write our the linear regression equation that corresponds to this fitted model. Make sure that you use the appropriate notation discussed in class.

## 7. Model Selection

### 7.1. Best Model

Out of all the models that we have tested with cross-validation in this analysis, which one would you select to predict EU_Sales for *new videos games*?

### 7.2. Actual New Dataset Performance

If we were to use this best selected model to predict EU_Sales for an actual new dataset, would you be suprised to see a model R^2 value of 0.35 for this new dataset? Why or why not?